keps/sig-storage/20191031-storage-capacity-constraints-for-pod-scheduling.md

```diff
@@ -0,0 +1,935 @@
+and higher fault tolerance by selecting striping respectively
+mirroring or raid in the storage class parameters of a driver for LVM,
+like for example [TopoLVM](https://github.com/cybozu-go/topolvm).
+
+The maximum size of the resulting volume then depends on the storage
+class and its parameters.
+
+#### Network attached storage
+
+In contrast to local storage, network attached storage can be made
+available on more than just one node. However, for technical reasons
+(high-speed network for data transfer inside a single data center) or
+political reasons (data must only be stored and processed in a single
+jurisdication) availability may still be limited to a subset of the
+nodes in a cluster.
+
+#### Custom schedulers
+
+For situations not handled by the Kubernetes scheduler now and/or in
+the future, a [scheduler
+extender](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/scheduling/scheduler_extender.md)
+can influence pod scheduling based on the information exposed via the
+new API. The
+[topolvm-scheduler](https://github.com/cybozu-go/topolvm/blob/master/docs/design.md#how-the-scheduler-extension-works)
+currently does that with a driver-specific way of storing capacity
+information.
+
+#### Operators for applications
+
+Application operators for modern scale out storage services
+(e.g. MongoDB, ElasticSearch, Kafka, MySQL, PostgreSQL, Minio, etc.)
+may want to control creation of volumes carefully in order to optimize
+availability, durability, performance and cost. For more information
+about this, see the [StoragePool API for Advanced Storage Placement
+KEP](https://github.com/kubernetes/enhancements/pull/1347).
+
+Storage pools as introduced in this KEP enable the creation of such
+operators by providing them the necessary information about storage in
+the cluster.
+
+### Caching remaining capacity via the API server
+
+CSI defines the `GetCapacity` RPC call for the controller service to
+report information about the current capacity but it's not accessible
+in the Kubernetes scheduler. A new type `CSIStoragePool` type gets
+added to store the result of the `GetCapacity` calls for one storage
+pool in the API server and thus make it available to the scheduler. In
+the future it might get extended to also store other information like
+health of the pool.
+
+### Identifying storage pools
+
+A method for identifying storage pools accessible via a CSI driver
+based on the existing topology information gets outlined below. The
+`external-provisioner` will be extended to handle the management of
+`CSIStoragePool` objects.
+
+For local storage in the node, there will be one such `CSIStoragePool`
+object per node and CSI driver on that node. For network attached
+storage, there will be one `CSIStoragePool` per CSI
+[Topology](https://github.com/container-storage-interface/spec/blob/4731db0e0bc53238b93850f43ab05d9355df0fd9/lib/go/csi/csi.pb.go#L1662-L1691)
+and driver.
+
+### Size of ephemeral inline volumes
```

Comment:
Can you add a section on how capacity reporting of ephemeral volumes will interact with the current ephemeral volumes storage allocatable feature, which reports based on the available rootfs size?

Is it expected that ephemeral drivers that do utilize the node boot disk do not use this feature? Otherwise you will have multiple sources reporting the same node boot disk capacity, potentially causing the node to be over committed. So that use case won't be solved by this.

Reply:
See 94729015b5dc96293f66ffdbc441902034a202d2.